clang/lib/Format/TokenAnnotator.cpp

```diff
@@ -4842,7 +4842,7 @@ bool TokenAnnotator::spaceRequiredBefore(const AnnotatedLine &Line,
     return true;
   }
   if (Left.is(TT_UnaryOperator)) {
-    if (Right.isNot(tok::l_paren)) {
+    if (!Right.isOneOf(tok::r_paren, tok::l_paren, tok::exclaim)) {
```

Comment:
> The only exception I see are identifiers and numeric literals (as they will combine the tokens together):
> 
> ```c++
> v(not x)
> v(not 1)
> v(not 123.f)
> ```

+1.

> Should this condition be something like `if (Right.isOneOf(tok::ident, /*... numeric literals*/))`? Or will this cause us to not add spaces in undesired places?
> 
> If we can avoid those spaces for only a limited set of tokens, I would probably limit this to `tok::lparen` and `tok::rparen`, the rest seems much less likely.

We should come up with a more general fix.

Reply:
For `not!`, I think adding a space is the right choice for a C++ code formatter: `not !b` should probably be formatted with a space, not as `not!b`, even though it's a pretty strange thing to write.

I wonder if this should instead be:
```suggestion
    if (!Right.isOneOf(tok::r_paren, tok::l_paren, tok::comma)) {
```
... with the `)` and `,` cases firing only in places where we can locally tell this isn't actually a `not` operator at all.

Reply:
Another data point: the [Python style guide](https://peps.python.org/pep-0008/#other-recommendations) also recommends always putting a space after `not`.

Reply:
Adding a space makes total sense in code that **has an argument** to `not`.
But this case seems special because `V(not!)` or `V(not)` clearly can't be a full expression by themselves. So it feels preferable to not add any extra spaces.